### PR TITLE
BUG: Series.groupby.size returning int64 for masked and arrow types

### PR DESCRIPTION
**Pull request by mroeschke**
- [ ] closes #xxxx (Replace xxxx with the GitHub issue number)
- [ ] [Tests added and passed](https://pandas.pydata.org/pandas-docs/dev/development/contributing_codebase.html#writing-tests) if fixing a bug or adding a new feature
- [ ] All [code checks passed](https://pandas.pydata.org/pandas-docs/dev/development/contributing_codebase.html#pre-commit).
- [ ] Added [type annotations](https://pandas.pydata.org/pandas-docs/dev/development/contributing_codebase.html#type-hints) to new arguments/methods/functions.
- [ ] Added an entry in the latest `doc/source/whatsnew/vX.X.X.rst` file if fixing a bug or adding a new feature.

### CODE REVIEW

File: pandas/core/groupby/groupby.py [hunk 1]
```diff
@@ -2890,12 +2891,21 @@ def size(self) -> DataFrame | Series:
         Freq: MS, dtype: int64
         """
         result = self.grouper.size()
+        result_dtype: str | np.dtype = result.dtype
+        if isinstance(self.obj, Series):
+            if isinstance(self.obj.array, ArrowExtensionArray):
+                result_dtype = "int64[pyarrow]"
+            elif isinstance(self.obj.array, BaseMaskedArray):
+                result_dtype = "Int64"
+        # TODO: For DataFrames what if columns are mixed arrow/numpy/masked?
```
Comment (rhshadrach) [inline pandas/core/groupby/groupby.py:2900]: Currently adding Series np + masked or np + arrow gives masked and arrow respectively (regardless of order). I think we could follow the same here. However masked + arrow raises whereas I would think we need to make a choice here. Not sure if this has been discussed much in other contexts?

File: pandas/tests/extension/test_arrow.py [hunk 2]
```diff
@@ -3121,6 +3121,13 @@ def test_iter_temporal(pa_type):
     assert result == expected
 
 
+def test_groupby_series_size_returns_pa_int(data):
+    ser = pd.Series(data[:3], index=["a", "a", "b"])
```
Comment (rhshadrach) [inline pandas/tests/extension/test_arrow.py:3125]: Can you add a GH ref

File: pandas/tests/groupby/test_size.py [hunk 3]
```diff
@@ -95,3 +95,11 @@ def test_size_on_categorical(as_index):
         expected = expected.set_index(["A", "B"])["size"].rename(None)
 
     tm.assert_equal(result, expected)
+
+
+@pytest.mark.parametrize("dtype", ["Int64", "Float64", "boolean"])
+def test_size_series_masked_type_returns_Int64(dtype):
+    ser = Series([1, 1, 1], index=["a", "a", "b"], dtype=dtype)
```
Comment (rhshadrach) [inline pandas/tests/groupby/test_size.py:102]: GH ref

**Review by rhshadrach — CHANGES_REQUESTED** [review]

File: pandas/core/groupby/groupby.py [hunk 4]
```diff
@@ -2890,12 +2891,21 @@ def size(self) -> DataFrame | Series:
         Freq: MS, dtype: int64
         """
         result = self.grouper.size()
+        result_dtype: str | np.dtype = result.dtype
+        if isinstance(self.obj, Series):
```
Comment (jbrockmendel) [inline pandas/core/groupby/groupby.py:2895]: can we just use convert_dtypes for this?  I think you've got another PR doing something similar for another method?

(i actually think for methods that return ints-and-never-NAs it is more useful to return non-nullable dtype, but whatever)

File: pandas/core/groupby/groupby.py (hunk 1 above)
Comment (mroeschke) [reply pandas/core/groupby/groupby.py:2900]: Yeah I don't think this has been discussed (but has been looming). I would prefer to punt on this for now until it's more formally discussed

File: pandas/core/groupby/groupby.py (hunk 4 above)
Comment (mroeschke) [reply pandas/core/groupby/groupby.py:2895]: > can we just use convert_dtypes for this?

Sure, I will still have to do this original-object introspection though to set the correct `dtype_backend` for `convert_dtype` if that's alright with you

File: pandas/core/groupby/groupby.py (hunk 1 above)
Comment (jbrockmendel) [reply pandas/core/groupby/groupby.py:2940]: im fine with punting.  id also point out that `__pandas_priority__` was implemented partly with this in mind

File: pandas/core/groupby/groupby.py (hunk 4 above)
Comment (jbrockmendel) [reply pandas/core/groupby/groupby.py:2935]: either way i guess

Comment (mroeschke) [reply pandas/core/groupby/groupby.py:2935]: Used `convert_dtypes` instead

File: pandas/tests/groupby/test_size.py (hunk 3 above)
Comment (mroeschke) [reply pandas/tests/groupby/test_size.py:103]: Added

**Review by rhshadrach — APPROVED** [review]
lgtm

**Pull request merged**